tracer/test/Datadog.Trace.Tests/Debugger/ConcurrentAdaptiveCacheTests.cs

```diff
@@ -0,0 +1,299 @@
+        }
+
+        [Fact]
+        public void GetOrAdd_Should_Add_Item_If_Not_Present()
+        {
+            var cache = new ConcurrentAdaptiveCache<string, int>(capacity: 10);
+
+            int result = cache.GetOrAdd("key", k => 42);
+            Assert.Equal(42, result);
+
+            Assert.True(cache.TryGet("key", out int value));
+            Assert.Equal(42, value);
+        }
+
+        [Fact]
+        public void GetOrAdd_Should_Return_Existing_Item_If_Present()
+        {
+            var cache = new ConcurrentAdaptiveCache<string, int>(capacity: 10);
+
+            cache.Add(42, keys: "key");
+            int result = cache.GetOrAdd("key", k => 100);
+            Assert.Equal(42, result);
+        }
+
+        [Fact]
+        public void Cache_Should_Evict_Items_When_Capacity_Is_Reached()
+        {
+            var cache = new ConcurrentAdaptiveCache<int, string>(capacity: 3, evictionPolicyKind: EvictionPolicy.LRU);
+
+            cache.Add("one", keys: 1);
+            cache.Add("two", keys: 2);
+            cache.Add("three", keys: 3);
+
+            // This should evict the least recently used item (1)
+            cache.Add("four", keys: 4);
+
+            Assert.False(cache.TryGet(1, out _));
+            Assert.True(cache.TryGet(2, out _));
+            Assert.True(cache.TryGet(3, out _));
+            Assert.True(cache.TryGet(4, out _));
+        }
+
+        [Fact]
+        public void Cache_Should_Use_LRU_Eviction_Policy_Correctly()
+        {
+            var cache = new ConcurrentAdaptiveCache<int, string>(capacity: 3, evictionPolicyKind: EvictionPolicy.LRU);
+
+            cache.Add("one", keys: 1);
+            cache.Add("two", keys: 2);
+            cache.Add("three", keys: 3);
+
+            // Access item 1, making it the most recently used
+            cache.TryGet(1, out _);
+
+            // This should evict item 2 (least recently used)
+            cache.Add("four", keys: 4);
+
+            Assert.True(cache.TryGet(1, out _));
+            Assert.False(cache.TryGet(2, out _));
+            Assert.True(cache.TryGet(3, out _));
+            Assert.True(cache.TryGet(4, out _));
+        }
+
+        [Fact]
+        public void Cache_Should_Use_LFU_Eviction_Policy_Correctly()
+        {
+            var cache = new ConcurrentAdaptiveCache<int, string>(capacity: 3, evictionPolicyKind: EvictionPolicy.LFU);
+
+            cache.Add("one", keys: 1);
+            cache.Add("two", keys: 2);
+            cache.Add("three", keys: 3);
+
+            // Access item 1 twice, making it the most frequently used
+            cache.TryGet(1, out _);
+            cache.TryGet(1, out _);
+
+            // Access item 2 once
+            cache.TryGet(2, out _);
+
+            // This should evict item 3 (least frequently used)
+            cache.Add("four", keys: 4);
+
+            Assert.True(cache.TryGet(1, out _));
+            Assert.True(cache.TryGet(2, out _));
+            Assert.False(cache.TryGet(3, out _));
```

Comment:
Does this access evict item 4 since that is the least used out of 1, 2, and 4?

Reply:
I'm not sure I understood the question so let me explain and correct me if I missed something. The goal of the test is to verify that the eviction policy is correct in case of LFU and capacity of 3. In this case we expect to see 1, 2 and 4 because 3 is the least used out of 1 and 2 (4 is the new entry).

Reply:
Got it, this all makes sense. I had a small misunderstanding and thought that they `cache.TryGet(3, out_)` call might add to the cache but I realize that is not the case. Thanks!